vmm/src/device_manager.rs

```diff
@@ -2924,12 +2919,37 @@ impl Snapshottable for DeviceManager {
         }
 
         // Then restore all devices associated with the DeviceManager.
-        for (id, dev) in self.migratable_devices.iter() {
-            debug!("Restoring {} from DeviceManager", id);
-            if let Some(snapshot) = snapshot.snapshots.get(id) {
-                dev.lock().unwrap().restore(*snapshot.clone())?;
-            } else {
-                return Err(MigratableError::Restore(anyhow!("Missing device {}", id)));
+        // It's important to restore devices in the right order, that's why
+        // the device tree is the right way to ensure we restore a child before
+        // its parent node.
+        for (id, node) in self.device_tree.iter() {
+            if node.child.is_some() {
+                continue;
+            }
+
+            // Restore the node
+            if let Some(device) = self.migratable_devices.get(id) {
+                debug!("Restoring {} from DeviceManager", id);
+                if let Some(snapshot) = snapshot.snapshots.get(id) {
+                    device.lock().unwrap().restore(*snapshot.clone())?;
+                } else {
+                    return Err(MigratableError::Restore(anyhow!("Missing device {}", id)));
+                }
+            }
+
```

Comment:
As a follow-up improvement, I think we want to have each and every of our device to implement the `Migratable` trait and the device tree `Node` to carry the device as `dyn Migratable` and unconditionally restore each device. Those who are practically not migratable will have an empty implementation for that trait.

For now, we could easily have an `migratable: Option<Arc<Mutex<dyn Migratable>>>>` added to our `Node`, and use our device tree instead of the `migratable_devices` map.